File: frameworks/micronaut-graphql/src/main/java/com/velocitybench/controller/HealthController.java
package com.velocitybench.controller;

import io.micronaut.http.HttpResponse;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

import javax.sql.DataSource;
import java.sql.Connection;
import java.util.Map;

@Controller
public class HealthController {

    private final DataSource dataSource;

    public HealthController(DataSource dataSource) {
        this.dataSource = dataSource;
    }

    @Get("/health")
    @Produces(MediaType.APPLICATION_JSON)
    public HttpResponse<Map<String, String>> health() {
        boolean healthy = checkDatabase();
        Map<String, String> response = Map.of(
            "status", healthy ? "healthy" : "unhealthy",
            "framework", "micronaut-graphql"
        );
        return healthy ? HttpResponse.ok(response) : HttpResponse.serverError(response);
    }

    @Get("/metrics")
    @Produces(MediaType.TEXT_PLAIN)
    public String metrics() {
        return """
            # HELP micronaut_requests_total Total number of GraphQL requests
            # TYPE micronaut_requests_total counter
            micronaut_requests_total 0
            # HELP micronaut_db_pool_size Database connection pool size
            # TYPE micronaut_db_pool_size gauge
            micronaut_db_pool_size 50
            """;
    }

    private boolean checkDatabase() {
        try (Connection conn = dataSource.getConnection()) {
            return conn.isValid(5);
        } catch (Exception e) {
            return false;
        }
    }
}

File: frameworks/play-graphql/app/controllers/HealthController.scala
package controllers

import db.Database
import play.api.libs.json._
import play.api.mvc._

import javax.inject._

@Singleton
class HealthController @Inject()(
  cc: ControllerComponents,
  database: Database
) extends AbstractController(cc) {

  def health: Action[AnyContent] = Action {
    val healthy = database.isHealthy
    val response = Json.obj(
      "status" -> (if (healthy) "healthy" else "unhealthy"),
      "framework" -> "play-graphql"
    )
    if (healthy) Ok(response) else ServiceUnavailable(response)
  }

  def metrics: Action[AnyContent] = Action {
    val metricsText =
      s"""# HELP play_requests_total Total number of GraphQL requests
         |# TYPE play_requests_total counter
         |play_requests_total 0
         |# HELP play_db_pool_size Database connection pool size
         |# TYPE play_db_pool_size gauge
         |play_db_pool_size ${database.poolSize}
         |""".stripMargin

    Ok(metricsText).as("text/plain")
  }
}
